sql-plugin/src/main/java/com/nvidia/spark/rapids/HashedPriorityQueue.java

```diff
@@ -77,14 +77,20 @@ public int size() {
 
   @Override
   public boolean offer(T obj) {
-    ensureCapacityToInsert();
-    // Start with the new object at the bottom of the heap and sift it up
-    // until heap properties are restored.
-    MutableInt location = new MutableInt(size);
-    locationMap.put(obj, location);
-    size += 1;
-    siftUp(obj, location);
-    return true;
+    if (!locationMap.containsKey(obj)) {
+      ensureCapacityToInsert();
+      // Start with the new object at the bottom of the heap and sift it up
+      // until heap properties are restored.
+      MutableInt location = new MutableInt(size);
+      locationMap.put(obj, location);
+      size += 1;
+      siftUp(obj, location);
+      return true;
```

Comment:
This does a wasteful double-lookup on the obj value.  This can be made more efficient by letting the `put` tell us if it's already there, e.g.:
```suggestion
    ensureCapacityToInsert();
    MutableInt location = new MutableInt(size);
    val old = locationMap.putIfAbsent(obj, location);
    if (old == null) {
      // Start with the new object at the bottom of the heap and sift it up
      // until heap properties are restored.
      size += 1;
      siftUp(obj, location);
      return true;
    } else {
      return false;
    }
```

Reply:
Much better, thank you!

Reply:
`ensureCapacityToInsert` may grow the array when it doesn't need to. Perhaps another approach is to attempt `putIfAbsent` and if we are successful, and then OOM, we can remove the location from `locationMap` to bring it back to a good state before re-throwing the OOM.

Reply:
ah this call should be a noop in the case I described.